Source/WebCore/Modules/url-pattern/URLPatternParser.cpp

```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+}
+
+// https://urlpattern.spec.whatwg.org/#generate-a-pattern-string
+String generatePatternString(const Vector<Part>& partList, const URLPatternStringOptions& options)
+{
+    StringBuilder result;
+
+    for (size_t index = 0; index < partList.size(); ++index) {
+        const Part& part = partList[index];
+
+        std::optional<Part> previousPart;
+        if (index > 0)
+            previousPart = partList[index - 1];
+
+        std::optional<Part> nextPart;
+        if (index < partList.size() - 1)
+            nextPart = partList[index + 1];
+
+        if (part.type == PartType::FixedText) {
+            if (part.modifier == Modifier::None) {
+                result.append(escapePatternString(part.value));
+
+                continue;
+            }
+            result.append(makeString("{"_s, escapePatternString(part.value), "}"_s, convertModifierToString(part.modifier)));
+
+            continue;
+        }
+
+        bool customName = !part.name.isEmpty() && !isASCIIDigit(part.name[0]);
+
+        bool needsGrouping = !part.suffix.isEmpty() || (!part.prefix.isEmpty() && part.prefix != options.prefixCodepoint);
+
+        if (!needsGrouping && customName
+            && part.type == PartType::SegmentWildcard && part.modifier == Modifier::None
+            && nextPart && nextPart->prefix.isEmpty() && nextPart->suffix.isEmpty()) {
+            if (nextPart->type == PartType::FixedText) {
+                // FIXME: Figure out if should be true or  false.
+                needsGrouping = isValidNameCodepoint(*StringView(nextPart->value).codePoints().begin(), true);
+            } else
+                needsGrouping = isASCIIDigit(nextPart->name[0]);
+        }
+
+        if (!needsGrouping && part.prefix.isEmpty() && previousPart && previousPart->type == PartType::FixedText) {
+            if (options.prefixCodepoint.length() == 1
+                && options.prefixCodepoint.startsWith(*StringView(nextPart->value).codePoints().codePointAt(nextPart->value.length() - 1)))
+                needsGrouping = true;
+        }
+
+        ASSERT(!part.name.isEmpty());
+
+        if (needsGrouping)
+            result.append("{"_s);
+
+        result.append(escapePatternString(part.prefix));
+
+        if (customName)
+            result.append(makeString(":"_s, part.name));
+
+        if (part.type == PartType::Regexp)
+            result.append("("_s, part.value, ")"_s);
+        else if (part.type == PartType::SegmentWildcard && !customName)
+            result.append(makeString("("_s, generateSegmentWildcardRegexp(options), ")"_s));
+        else if (part.type == PartType::FullWildcard) {
+            if (!customName
+                && (!previousPart || previousPart->type == PartType::FixedText || previousPart->modifier != Modifier::None
+                    || needsGrouping || !part.prefix.isEmpty()))
+                result.append("*"_s);
+            else
+                result.append(makeString("("_s, ".*"_s,  ")"_s));
```

Comment:
Seems like the least efficient way to write:
```
result.append("(.*)"_s);
```

Reply:
Yup! My bad, the spec confused me a little bit with this. This has been fixed. 
<img width="579" alt="Screenshot 2024-11-15 at 6 09 03 PM" src="https://github.com/user-attachments/assets/2be45bff-3646-4705-84ab-81a9c341e4d8">